### ISSUE TITLE
remove unimplemented cf_app bluegreen support

### ISSUE THREAD

**Issue by gberche-orange** [cloudfoundry-community/terraform-provider-cf#62]
The cf_app.route is designed to support blue/green deployment but is not currently implemented. 

See https://github.com/mevansam/terraform-provider-cf/blob/master/website/docs/r/app.html.markdown#routing-and-blue-green-deployment-strategy

>     TO BE IMPLEMENTED
> 
>     stage_route - (Optional, String) The GUID of the route where the staged application will be available.
>     live_route - (Optional, String) The GUID of the route where the live application will be available.
>     validation_script - (Optional, String) The validation script to execute against the stage application before mapping the live route to the staged application.
> 

Related issue is #25

This should be removed from documentation and from resource meta-data as to avoid breaking changes post 1.0 when the support gets implemented (or zero deployment support gets added instead)

As a consequence typing of the `cf_app.route` can simplified to remove unnecessary nesting

> * `route` - (Optional) Configures how the application will be accessed externally to cloudfoundry. 
>   - `default_route` - (Optional, String) The GUID of the default route where the application will be available once deployed.

**Comment by gberche-orange** [comment]
related email exchange with @jcarrothers-sap 

> I was hoping to get some early feedback on some changes I’ve been working on related to route handling in the app resource.  I’ve been working on blue/green support within the provider for the last week and I’ve come to the conclusion that the current route handling needs a complete revamp.  This is the same revamp that I posted about in the slack channel about on Tuesday.  
> Please find the complete set of preliminary, untested changes here: https://github.com/jcarrothers-sap/terraform-provider-cf/compare/ebfd1d9bfdce2f5bdd924ff89b1762d2cd691d3d...fixRouteUpdates_rework
> 
> The shorter version is that I’m intending to replace the current “route” structure with something along the lines of the following:
> 
> "routes": &schema.Schema{
>     Type:          schema.TypeSet,
>     Optional:      true,
>     MinItems:      1,
>     ConflictsWith: []string{"route"},
>     Set:           hashRouteMappingSet,
>     Elem: &schema.Resource{
>         Schema: map[string]*schema.Schema{
>             "route": &schema.Schema{
>                 Type:         schema.TypeString,
>                 Required:     true,
>                 ValidateFunc: validation.NoZeroValues,
>             },
>             "port": &schema.Schema{
>                 Type:         schema.TypeInt,
>                 Optional:     true,
>                 Computed:     true,
>                 Deprecated:   "Not yet implemented!",
>                 ValidateFunc: validation.IntBetween(1, 65535),
>             },
>             "mapping_id": &schema.Schema{
>                 Type:     schema.TypeString,
>                 Computed: true,
>             },
>             "exclusive": &schema.Schema{
>                 Type:        schema.TypeBool,
>                 Description: "Should terraform remove all mappings of this route not declared here?",
>                 Deprecated:  "Not yet implemented!",
>                 Optional:    true,
>                 Default:     true,
>             },
>         },
>     },
> },
> 

```

**Comment by gberche-orange** [comment]
@jcarrothers-sap 

Please consider submitting your changes as a distinct pr on the `mevansam/terraform-provider-cf` repo as this will help us use github codereview features as well as the travis-ci test results.

In the meantime, here is some comments inline 

> * `routes` - (Optional, Array) Configures how the application will be accessed externally to cloudfoundry. 
> 
>   - `route` - (Required, String) The route GUID.
>   - `port` - (Optional, Number) The port of the application that the route mapping is to bind to (NOT YET IMPLEMENTED!)
>   - `exclusive` - (Option, Default: true) When true, terraform will enforce that the route is not to have any other mappings not controlled by this resource (Note: "false" value NOT YET IMPLEMENTED!)

+1 for the renaming of `route.default_route` into `routes.route` 

Regarding the `port parameter` we should add support for tcp routes in a distinct issue, and document it as this time. Tracking this as #115

Regarding the `exclusive` parameter, given that the [route_mappings](http://apidocs.cloudfoundry.org/2.3.0/routes_mapping/mapping_an_app_and_a_route.html) objects is provisionned and owned by the terraform provider, I don't believe the provider should support cases with `exclusive=false`. I would therefore advocate against this parameter.

**Comment by jcarrothers-sap** [comment]
@gberche-orange 

The `port` parameter isn't intended to be about TCP routes, unless I'm misunderstanding CF's route architecture.  I was specifically thinking that we should eventually implement support for the `app_port` parameter in this API: https://apidocs.cloudfoundry.org/2.3.0/routes_mapping/mapping_an_app_and_a_route.html

I've fine with removing the exclusive option from these changes.  I included it because of an idea I was toying with and because, as a side benefit, including it in the documentation explicitly calls out this behaviour in the app resource.

**Comment by gberche-orange** [comment]
@jcarrothers-sap my understanding is that the `port` param of the [route_mappings](https://apidocs.cloudfoundry.org/2.3.0/routes_mapping/mapping_an_app_and_a_route.html) endpoint is only used in the case of TCP routes, see CLI usage below which only mentions the `port` param for TCP routes

```
$ cf map-route --help
NAME:
   map-route - Add a url route to an app

USAGE:
   Map an HTTP route:
      cf map-route APP_NAME DOMAIN [--hostname HOSTNAME] [--path PATH]

   Map a TCP route:
      cf map-route APP_NAME DOMAIN (--port PORT | --random-port)

EXAMPLES:
[...]
   cf map-route my-app example.com --port 5000                  # example.com:5000
```

**Comment by jcarrothers-sap** [comment]
@gberche-orange My understanding is that, in this case, the documentation is referring to `cf map-route`'s capability to create the requested route if it doesn't already exist.  The CF API documentation makes no reference to this parameter being exclusively for TCP routes, and I don't see why they would have limited it like that anyways.

https://apidocs.cloudfoundry.org/1.40.0/routes_mapping/mapping_an_app_and_a_route.html

**Comment by gberche-orange** [comment]
@jcarrothers-sap AFAIK the apps binding http routes (i.e. not TCP routes nor internal DNS discovery routes) must listed on the port specified by the [PORT](https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#PORT) variable, and I'm not aware of any support for developers to change this port (which usually is 8080 on diego).

The `create-route` CLI command doesn't either export a port parameter except for TCP routes

```
cf create-route --help
NAME:
   create-route - Create a url route in a space for later use

USAGE:
   Create an HTTP route:
      cf create-route SPACE DOMAIN [--hostname HOSTNAME] [--path PATH]

   Create a TCP route:
      cf create-route SPACE DOMAIN (--port PORT | --random-port)

EXAMPLES:
   cf create-route my-space example.com                             # example.com
   cf create-route my-space example.com --hostname myapp            # myapp.example.com
   cf create-route my-space example.com --hostname myapp --path foo # myapp.example.com/foo
   cf create-route my-space example.com --port 5000                 # example.com:5000

OPTIONS:
   --hostname, -n      Hostname for the HTTP route (required for shared domains)
   --path              Path for the HTTP route
   --port              Port for the TCP route
   --random-port       Create a random port for the TCP route
```

**Comment by jcarrothers-sap** [comment]
@gberche-orange My understanding is that, in this case, the documentation is referring to `cf map-route`'s capability to create the requested route if it doesn't already exist.  The CF API documentation makes no reference to this parameter being exclusively for TCP routes, and I don't see why they would have limited it like that anyways.

https://apidocs.cloudfoundry.org/1.40.0/routes_mapping/mapping_an_app_and_a_route.html

**Comment by jcarrothers-sap** [comment]
This is correct, in order for you to specify the port an application listens on, you have to be using diego.  Furthermore, the API does specify that when you map a route, the target port must be one of the ports the application is listening on.  The default, when you don't specify the port (through the API), is to map the route to the first port listed by the application it is being bound to.

I agree, this is very much an edge case which is why it is not implemented yet, but I'm fairly certain the API supports this.  I will try and fine time to test this.

**Comment by gberche-orange** [comment]
Will be included with #150
